File: frontend/propmate-web/src/services/viewingService.js
const API_URL = "http://localhost:5235/api";

// Get available viewing slots for a property
export async function getViewingSlots(propertyId) {
    const response = await fetch(
        `${API_URL}/properties/${propertyId}/viewing-slots`
    );

    if (!response.ok) {
        throw new Error("Failed to load viewing slots.");
    }

    return await response.json();
}

// Create a new viewing slot
export async function createViewingSlot(
    propertyId,
    startTime,
    endTime
) {
    const response = await fetch(`${API_URL}/viewing-slots`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            propertyId,
            startTime,
            endTime,
        }),
    });

    if (!response.ok) {
        throw new Error("Failed to create viewing slot.");
    }

    return await response.json();
}

// Book a viewing slot
export async function bookViewing(viewingSlotId, userId) {
    const response = await fetch(`${API_URL}/viewing-bookings`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            viewingSlotId,
            userId,
        }),
    });

    if (!response.ok) {
        if (response.status === 409) {
            throw new Error("This viewing slot is already booked.");
        }

        throw new Error("Failed to book viewing.");
    }

    return await response.json();
}

// Cancel a viewing booking
export async function cancelViewing(bookingId) {
    const response = await fetch(
        `${API_URL}/viewing-bookings/${bookingId}`,
        {
            method: "DELETE",
        }
    );

    if (!response.ok) {
        throw new Error("Failed to cancel viewing.");
    }
}

File: mobile/propmate_mobile/lib/services/viewing_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/viewing_slot.dart';

class ViewingService {
  static const String baseUrl = 'http://localhost:5235/api';

  Future<List<ViewingSlot>> getViewingSlots(int propertyId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/properties/$propertyId/viewing-slots'),
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);

      return data.map((json) => ViewingSlot.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load viewing slots');
    }
  }
}
